### PR TITLE
Mania endtime object conversion

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x]  https://github.com/ppy/osu/pull/788
- [x] https://github.com/ppy/osu/pull/820
- [x] https://github.com/ppy/osu/pull/822

I think one at a time is the best for these, so here's the one that creates patterns for spinners.

### CODE REVIEW

File: osu.Game.Rulesets.Mania/Beatmaps/Patterns/Legacy/PatternType.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace osu.Game.Rulesets.Mania.Beatmaps.Patterns.Legacy
+{
+    /// <summary>
+    /// The type of pattern to generate. Used for legacy patterns.
+    /// </summary>
+    [Flags]
+    internal enum PatternType
+    {
+        None = 0,
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Beatmaps/Patterns/Legacy/PatternType.cs]: Can we use `= 1 << 0` notation instead?

File: osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        private readonly List<ManiaHitObject> hitObjects = new List<ManiaHitObject>();
+
+        /// <summary>
+        /// All the hit objects contained in this pattern.
+        /// </summary>
+        public IEnumerable<ManiaHitObject> HitObjects => hitObjects;
+
+        /// <summary>
+        /// Whether this pattern already contains a hit object in a code.
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs]: "Check whether" "column"

File: osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        /// </summary>
+        /// <param name="column">The column index.</param>
+        /// <returns>Whether this pattern already contains a hit object in <paramref name="column"/></returns>
+        public bool IsFilled(int column) => hitObjects.Exists(h => h.Column == column);
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs]: `CheckColumnHasObject` might be a better name.

File: osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        /// <summary>
+        /// Amount of columns taken up by hit objects in this pattern.
+        /// </summary>
+        public int ColumnsFilled => HitObjects.GroupBy(h => h.Column).Count();
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs]: `ColumnsWithObjects` maybe

File: osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        /// <param name="hitObject">The hit object to add.</param>
+        public void Add(ManiaHitObject hitObject) => hitObjects.Add(hitObject);
+
+        /// <summary>
+        /// Copies hit object from another pattern to this one.
+        /// </summary>
+        /// <param name="other">The other pattern.</param>
+        public void Add(Pattern other)
+        {
+            other.HitObjects.ForEach(Add);
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Beatmaps/Patterns/Pattern.cs]: might be able to use `hitObjects.AddRange(other.HitObjects)`

File: osu.Game.Rulesets.Mania/Objects/ManiaHitObject.cs [hunk 6]
```diff
@@ -9,5 +9,11 @@ namespace osu.Game.Rulesets.Mania.Objects
     public abstract class ManiaHitObject : HitObject, IHasColumn
     {
         public int Column { get; set; }
+
+        /// <summary>
+        /// The number of other <see cref="ManiaHitObject"/> that start at
+        /// the same time as this hit object.
+        /// </summary>
+        public int Siblings { get; set; }
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Objects/ManiaHitObject.cs]: o_O

File: osu.Game/Beatmaps/Events/EventInfo.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace osu.Game.Beatmaps.Events
+{
+    public class EventInfo
```
Comment (peppy) [inline osu.Game/Beatmaps/Events/EventInfo.cs]: still don't think we want this. just make a list of breaks inside Beatmap for now

**Review by peppy — CHANGES_REQUESTED** [review]

**Review by peppy — APPROVED** [review]

File: osu.Game.Rulesets.Mania/Beatmaps/Patterns/Legacy/EndTimeObjectPatternGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        /// <summary>
+        /// Picks a random column after a column.
+        /// </summary>
+        /// <param name="start">The starting column.</param>
+        /// <returns>A random column after <paramref name="start"/>.</returns>
+        private int getNextRandomColumn(int start)
+        {
+            int nextColumn = Random.Next(start, AvailableColumns);
+
+            while (PreviousPattern.IsFilled(nextColumn))
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Beatmaps/Patterns/Legacy/EndTimeObjectPatternGenerator.cs]: add a check whether all columns are filled before attempting this. could endless loop.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by smoogipooo** [comment]
Comments before the merge fixed in https://github.com/ppy/osu/pull/822 .

**Review by peppy — APPROVED** [review]

**Pull request merged**